Shared transitions

stagg opened the pull request:
## Shared Element API
- Creating two base elements for using shared transitions `SharedElementTransitionLayout` and `SharedElementTransitionScope`
- This is done to be able to indirectly provide a `SharedTransitionScope` to child elements (ie Screen) without having to pipe the scope all the way down into Circuit Ui's
- Circuit foundation is updated to provide a _Navigation_ scoped `AnimatedVisibilityScope` via a new `AnimatedNavDecoration` variant of `NavDecoration`
- Circuit overlays were also updated to provide an _Overlay_ `AnimatedVisibilityScope`

### `SharedElementTransitionScope`

- Provides a `SharedTransitionScope` for the standard shared elements/shared bounds animations 
- The `SharedElementTransitionScope` instance can then be access via the composable `SharedElementTransitionScope` 
- The `SharedElementTransitionScope` composable should be used like `SharedTransitionScope` to access the `sharedElement` and `sharedBound` modifiers
- Based on where/how the shared element will be shared `getAnimatedScope` or `requireAnimatedScope` can be used to access a needed `AnimatedVisibilityScope`
- `AnimatedVisibilityScope`s can be provided with the `ProvideAnimatedTransitionScope`  composable
- By default Circuit provides Navigation and Overlay scopes when a `SharedElementTransitionScope` is available

### `SharedElementTransitionLayout`
- `SharedElementTransitionLayout` is a layout that creates and provides a `SharedElementTransitionScope`
- Normally `SharedElementTransitionLayout` should be setup around the a standard circuit setup, between `CircuitCompositionLocals` and before `NavigableCircuitContent` or `ContentWithOverlays` 
- A `PreviewSharedElementTransitionLayout` is also provided to help with `@Preview` use


## Update star to use SharedElementTransitions
- Wired the shared elements into Star as a demonstration for the Navigation and Overlay use cases
- Gesture predictive back (for Android) has been updated to use a seekable transition to drive shared element transitions

|Scenario|Demo|
|-|-|
|No shared elements|<video src="https://github.com/user-attachments/assets/61fd57da-2090-4953-83ec-9573a9ed33d5"></video>|
|Navigation Scope|<video src="https://github.com/user-attachments/assets/de0b2b3b-73dc-40a7-ad81-b22d312ac535"></video>|
|Overlay Scope|<video src="https://github.com/user-attachments/assets/516a56ce-a8a7-4458-8b48-2ebcc4e925df"></video>|
|Seekable Predictive back|<video src="https://github.com/user-attachments/assets/8ba2b3a5-f091-436f-b1f3-95a6d0de217b"></video>|

ZacSweers commented:
Some high level early thoughts
- Can we add a control to disable the default nav decoration? This way sharedBounds() could be used without colliding with it. I'm almost wondering if we should disable the nav decoration by default if a shared element is present
- Related from the above, can we leverage `sharedBounds()` in one of the samples?
- Overlay API definitely seems like something we should get figured out with this

chrisbanes commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/internal/NoOpSharedTransitionScope.kt:
```diff
@@ -0,0 +1,104 @@
+      layoutDirection: LayoutDirection,
+      density: Density,
+    ): Path {
+      path.reset()
```
Use `rewind()` it's cheaper for the next outline

chrisbanes commented:
Do you have any sense of how much work is involved? I'd like to see a bit more progress before landing this personally, as I think there's going to be tests which are broken. 

If only just getting the sample looking right to prove it out fully.

riggaroo commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/internal/NoOpSharedTransitionScope.kt:
```diff
@@ -0,0 +1,104 @@
+ */
+@InternalCircuitApi
+@OptIn(ExperimentalSharedTransitionApi::class)
+public data object NoOpSharedTransitionScope : SharedTransitionScope {
```
Curious on the need for this, vs just not adding the modifiers / layout when not required. Hard to really understand all the details of Circuit but maybe you've explored that option already?

stagg replied:
Was initially experimenting with this from the view `Screen` reuse regardless of the context. Thinking on it some more its not really needed, a `Screen` really shouldn't be getting reused as such.

riggaroo commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/SharedElementTransition.kt:
```diff
@@ -0,0 +1,165 @@
+ * [SharedElementTransitionScope].
+ */
+@Composable
+private fun staticAnimatedVisibilityScope(): AnimatedVisibilityScope {
```
This feels like it could potentially introduce some UI inconsistencies, is it possible that a developer somehow relies on the scope to perform a different UI animations unrelated to shared elements? Maybe it'd be less error-prone to rather have a nullable scope instead. 

Or have the Nav framework itself always expose the AnimatedVisibilityScope and throw an exception if its not there as something would've gone quite wrong to have that happen?

doris4lt replied:
I second what Rebecca said. This custom scope will always have the target visibility be `EnterExitState.Visible`. That would potentially prevent the shared elements from identifying the target bounds. More specifically when both sharedElement/sharedBounds call sites of the same key report their target visibility to be visible, we would not know which bounds to use as the target bounds since they are both "entering".

stagg replied:
Yup, it's pretty broken always being in the single state right now. Going to lean on the nav for providing this.

riggaroo commented:
Love the visuals to aid the PR description! Thanks for showcasing them 🎉 

The animations are looking great too - love the back from full screen to the pager. 

Could the Active Scope (layout.mp4) example at 0:07 be improved to have the image shared to that screen too? It seems like there is a blank screen at that point. Maybe its an image loading issue, perhaps it could be corrected by setting a placeholder cache key to the same as the memory cache key? `.placeholderMemoryCacheKey("image-key") ` and having the placeholder be null, similar to this example https://developer.android.com/develop/ui/compose/animation/shared-elements/common-use-cases#async-images

From details -> list of dogs, the final clipping is applied immediately to the dog, consider animating the rounded corners too as it progresses between the states, the bottom corners can animation from rounded to straight - something similar was done here - https://youtu.be/PR6rz1QUkAM?si=DsGg4huAxgbel9c_&t=989

I do agree about getting predictive back working correctly too - as this may heavily influence your API design more, as developers would probably want to tie into the Transition object to do other surrounding animations, as well as it can help with debugging to be able to seek between the states.

stagg commented on samples/star/src/commonMain/kotlin/com/slack/circuit/star/petdetail/PetPhotoCarouselScreen.kt:
```diff
@@ -118,88 +125,96 @@ internal object PetPhotoCarouselTestConstants {
+        name = state.name,
+        photoUrlMemoryCacheKey = state.photoUrlMemoryCacheKey,
+        modifier =
+          Modifier.sharedElementWithCallerManagedVisibility(
```
This case is a bit different as it's dealing with an element being shared across navigation and again as an Overlay is being shown. Need to really work on the Overlay API for this a bit more, currently missing a `AnimatedVisibilityScope` from it 😅 Think the goal would be to dynamically switch to the correct `AnimatedVisibilityScope` based on whats happening.

Thanks for the `SeekableTransitionState` snippet 🙇🏻 The predictive back implementation/api is definitely going to need to provide some of the progress information.

stagg commented:
> * Can we add a control to disable the default nav decoration? This way sharedBounds() could be used without colliding with it. I'm almost wondering if we should disable the nav decoration by default if a shared element is present

We're going to need some form of `AnimatedVisibilityScope` from the nav decoration to run the transitions, so would this set the animation to something predefined?  

> Do you have any sense of how much work is involved? I'd like to see a bit more progress before landing this personally, as I think there's going to be tests which are broken.

Think most of the work is in supporting seekable transitions for predictive back, then maybe building some simpler api's on top of that?

chrisbanes commented:
> Think most of the work is in supporting seekable transitions for predictive back, then maybe building some simpler api's on top of that?

Sounds good to me! I just don't see any tests being ran on CI?

stagg commented:
Made some progress with the overlay transitions, bit more to tidy up with it and then on to predictive back support. Think that will involve some `NavDecoration` rework if we want to control the animation scope setup better.

https://github.com/user-attachments/assets/5b5a00ae-7007-48f0-9b28-c73c1756458e

ZacSweers commented:
Looking better. Def game to make whatever changes we need to nav decoration APIs (and overlays)

ZacSweers commented:
Some early thoughts from going around the sample before looking through the code
- Is there a reason it never works the first time in debug? It seems to only work the second+ time. On release builds it's fine

- Anything we can do to improve the transition for the photo viewer?

[starsampleone.webm](https://github.com/user-attachments/assets/0981e2b3-f0ca-4a2a-9941-4a9d77b1df8d)

- Do you think it would be possible to, in the photo viewer when it's dragged but not flung, do a shared element transition instead?

- The transition from the home list to the detail view feels asymmetric. When going from A -> B, the background content translates away. When going from B -> A, some content translates but the bottom nav bar translates up

- I notice when seeking through predictive back that the top of the stack exerts a significant shadow, which I think is causing some of the artifacting I'm seeing in the above bits. Anything we should tweak in NavigableCircuitContent or elsewhere to improve this? What's particularly weird is that it appears behind some content on the previous screen and separate from the top screen too.

[starsample2.webm](https://github.com/user-attachments/assets/ee26ef99-bbf2-4d8e-89d9-da9c7501c694)

- The sample crashes when I go to detail and rotate 😬. Doesn't crash on just the home screen.

- For the card, I wonder if we can animate its bounds out via shrinking rather than a full-screen expand. It looks a little like it blows out/flashes.

- If the photo carousel is scrolled in the detail screen, the shared element transition still runs but it blips back to the first image at the start.

[starsample3.webm](https://github.com/user-attachments/assets/0102b865-2ebf-4450-8275-4eba821c182a)

stagg commented:
> * Is there a reason it never works the first time in debug? It seems to only work the second+ time. On release builds it's fine

Not sure I've run into this? 

> * Anything we can do to improve the transition for the photo viewer?

Changed it a bit with 0d1b246

> * Do you think it would be possible to, in the photo viewer when it's dragged but not flung, do a shared element transition instead?

Nice idea, added with e972508

> * I notice when seeking through predictive back that the top of the stack exerts a significant shadow

Fixed with 2779634, the drawbacks of dark mode 😅

> * For the card, I wonder if we can animate its color quicker. It looks a little like it blows out/flashes.
> * If the photo carousel is scrolled in the detail screen, the shared element transition still runs but it blips back to the first image at the start.

Changed the animation timings for both of these. Jumps directly to the first image which fades in, but now won't do an odd cross fade.

> * The transition from the home list to the detail view feels asymmetric. When going from A -> B, the background content translates away. When going from B -> A, some content translates but the bottom nav bar translates up

Yaaa this has been the main problem. We don't know if there are shared transitions in time to disable the animated content transition. 
We can either have the screens tell us there's going to be a shared animation, which is some of what I was playing with using `AnimatedScreen`. That or I think we'd need to rebuild `AnimatedContent` in some way to let us compose both screens ahead of time. Which lets us check for `SharedTransitionScope.isTransitionActive`, and then build the transition spec 😬 

I like the idea of the `AnimatedScreen`/ `AnimatedNavigationOverride` as it'd dovetail into a more capable API for changing the enter/exit animation based on the screens involved in the transition. Think that should be a separate PR though.

ZacSweers commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/ModifierExt.kt:
Before you merge let's push these to `com.slack.circuit.foundation.internal`

ZacSweers commented on backstack/src/commonMain/kotlin/com/slack/circuit/backstack/Navigation.kt:
```diff
@@ -18,16 +18,22 @@ package com.slack.circuit.backstack
 }
+
+/** Argument provided to [NavDecoration] that exposes the underlying [Screen]. */
+public interface NavArgument {
```
what's the difference between this and accessing the screen via the `args` list (of records) that's passed in?

stagg replied:
Outside the specific implementation of `NavDecoration` we won't be able to access the specific screen as it'd end up being visible as `T`. Its needed for any animation override apis, not really for the current shared elements wok.

ZacSweers commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/NavigationDecoration.kt:
```diff
@@ -0,0 +1,91 @@
+import kotlinx.collections.immutable.ImmutableList
+
+@OptIn(ExperimentalSharedTransitionApi::class)
+public abstract class AnimatedNavDecoration(
```
Let's add a big ol' doc on this and its usage on this

ZacSweers commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/NavigationDecoration.kt:
```diff
@@ -0,0 +1,91 @@
+      decoratorFactory.create<T>() as AnimatedNavDecorator<T, AnimatedNavState>
+    }
+    with(decorator) {
+      Content(args, backStackDepth, modifier) { modifier ->
+        AnimatedContent(modifier = modifier, transitionSpec = transitionSpec()) { targetState ->
+          ProvideAnimatedTransitionScope(Navigation, this) {
+            AnimatedNavContent(targetState) { content(it) }
```
it's hard for me to look at this level of nesting and feel like these new compose animation APIs are supposed to be easy :(

ZacSweers commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/NavigationDecoration.kt:
```diff
@@ -0,0 +1,91 @@
+}
+
+@Stable
+public interface AnimatedNavDecorator<T : NavArgument, S : AnimatedNavState> {
```
same here re: doc

ZacSweers commented on circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/NavigationDecoration.kt:
```diff
@@ -0,0 +1,91 @@
+  )
+
+  @Stable
+  public interface Factory {
```
nit: `fun interface`?

ZacSweers commented on circuit-shared-elements/src/commonMain/kotlin/com/slack/circuit/sharedelements/AnimatedVisibilityScopeExt.kt:
```diff
@@ -0,0 +1,26 @@
+import androidx.compose.runtime.derivedStateOf
+
+/** Current progress fraction of the animation, between 0f and 1f. */
+public fun AnimatedVisibilityScope.progress(): FloatState {
```
nit: internal package?

ZacSweers commented on circuit-shared-elements/src/commonMain/kotlin/com/slack/circuit/sharedelements/DelicateCircuitSharedElementsApi.kt:
```diff
@@ -0,0 +1,12 @@
+@RequiresOptIn
+@Retention(AnnotationRetention.BINARY)
+@Target(FUNCTION, PROPERTY)
+public annotation class DelicateCircuitSharedElementsApi
```
Let's go the whole way and make this just `ExperimentalCircuitSharedElementsApi` for now

ZacSweers commented on circuit-shared-elements/src/commonMain/kotlin/com/slack/circuit/sharedelements/SharedElementTransitionLayout.kt:
```diff
@@ -0,0 +1,106 @@
+import com.slack.circuit.sharedelements.SharedElementTransitionScope.AnimatedScope.Navigation
+
+/**
+ * [SharedElementTransitionLayout] creates a layout with a [SharedElementTransitionScope]. Any child
```
let's add a simple sample snippet to this kdoc for laymen like me reading it

ZacSweers commented on circuit-shared-elements/src/commonMain/kotlin/com/slack/circuit/sharedelements/SharedElementTransitionScope.kt:
```diff
@@ -0,0 +1,177 @@
+  public fun availableScopes(): Set<AnimatedScope>
+
+  /** Get the [AnimatedVisibilityScope] for the given [key], if it exists. */
+  public fun getAnimatedScope(key: AnimatedScope): AnimatedVisibilityScope?
```
nit: `find` since it returns nullable?

ZacSweers commented on circuit-shared-elements/src/commonMain/kotlin/com/slack/circuit/sharedelements/SharedElementTransitionScope.kt:
```diff
@@ -0,0 +1,177 @@
+
+  /** Require the [AnimatedVisibilityScope] for the given [key]. Throwing if it does not. */
+  public fun requireAnimatedScope(key: AnimatedScope): AnimatedVisibilityScope {
+    return requireNotNull(getAnimatedScope(key)) { "No AnimatedVisibilityScope found for $key" }
```
nit: should actually be `checkNotNull`. require will throw an illegalargumentexception, which I suspect isn't quite semantically what you're aiming for here

ZacSweers commented on circuit-shared-elements/src/commonMain/kotlin/com/slack/circuit/sharedelements/SharedElementTransitionScope.kt:
```diff
@@ -0,0 +1,177 @@
+
+    /** Helper to check if a [SharedElementTransitionScope] has an active transition. */
+    @Composable
+    public fun isTransitionActive(): Boolean {
```
nit: composable property?

ZacSweers commented on docs/shared-elements.md:
💯

ZacSweers commented on docs/shared-elements.md:
````diff
@@ -0,0 +1,119 @@
+A `PreviewSharedElementTransitionLayout` is provided for use when using [`@Preview`](https://developer.android.com/develop/ui/compose/tooling/previews) for any Composable that has a `SharedElementTransitionScope`.
+
+```kotlin
+
````
```suggestion
```

ZacSweers commented on docs/shared-elements.md:
````diff
@@ -0,0 +1,119 @@
+  }
+
+```
+
````
```suggestion
```

ZacSweers commented on docs/shared-elements.md:
```diff
@@ -0,0 +1,119 @@
+
+### AnimatedScope
+
+By default Circuit provides a _Navigation_ and _Overlay_ `AnimatedVisibilityScope` when a `SharedElementTransitionScope` is available and either `NavigableCircuitContent` or `ContentWithOverlays` has been used. 
```
These are code refs right?

```suggestion
By default Circuit provides a `Navigation` and `Overlay` `AnimatedVisibilityScope` when a `SharedElementTransitionScope` is available and either `NavigableCircuitContent` or `ContentWithOverlays` has been used. 
```

ZacSweers commented on docs/shared-elements.md:
```diff
@@ -0,0 +1,119 @@
+    AnimatedNavContent(targetState) { content(it) }
+  }
+}
+
```
```suggestion
```

ZacSweers commented on docs/shared-elements.md:
```diff
@@ -0,0 +1,119 @@
+## Non-Circuit usage
+
+Circuit Shared Elements should be usable without any other Circuit artifact. The setup of `SharedElementTransitionLayout` is the same outside the root content, but a non-Circuit setup would require explicit use of `ProvideAnimatedTransitionScope` to provide the desired `AnimatedVisibilityScope`.
+
+
+
+
```
```suggestion

```

ZacSweers commented on samples/star/src/androidMain/kotlin/com/slack/circuit/star/MainActivity.kt:
```diff
@@ -60,7 +62,7 @@ class MainActivity @Inject constructor(private val circuit: Circuit) : AppCompat
         val httpUrl = intent.data.toString().toHttpUrl()
         val animalId = httpUrl.pathSegments[1].substringAfterLast("-").toLong()
-        val petDetailScreen = PetDetailScreen(animalId, null)
+        val petDetailScreen = PetDetailScreen(animalId, null, null)
```
nit: let's name these args now

ZacSweers commented on samples/star/src/commonMain/kotlin/com/slack/circuit/star/transition/PetSharedTransitionKeys.kt:
```diff
@@ -0,0 +1,17 @@
+// SPDX-License-Identifier: Apache-2.0
+package com.slack.circuit.star.transition
+
+interface SharedTransitionKey
```
do you think this is something worth encouraging in docs/promoting the marker interface to a shipped artifact?

ZacSweers reviewed (approved):
left some questions for my understanding in the PR! Can merge and answer them async. Some also here

- do we still need the separate artifact now that these are in compose-multiplatform?

---

All the comments non-blocking or nits, def think there's a lot here and we should roll forward and iterate more in followup PRs. Awesome work!
